### PR TITLE
Mvc implementation

### PR DESCRIPTION
**Pull request by denysLavrienov**
This PR implements MVC architecture.

Created:
* `User` - stands for a simple user, containing login, password, and avatar;
* `UserFile` - stands for uploaded user file;
* `LoginController` - creates `login form` and operates it;
* `RegistrationController` - creates `registration form` and operates it;
* `RequestService` - sends `POST` and `GET` requests to BE.

### CODE REVIEW

File: app/component/button.js [hunk 1]
```diff
@@ -31,4 +31,11 @@ export class Button {
   onClick(handler) {
     this.button.addEventListener('click', handler);
   }
+
+  /**
+   * Disables button after one click to prevent user`s multiple clicks.
+   */
+  disable() {
```
Comment (dkalinin) [inline app/component/button.js]: Does it have 'enable' method?
Maybe it's simpler have a `disabled` boolean field.

File: app/component/link.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    this.URL = URL;
+    this.render();
+  }
+
+  /**
+   * Renders and displays link object.
+   */
+  render() {
+    const link = document.createElement('span');
+    link.innerHTML = `<br><a href = ${this.URL}>${this.text}</a>`;
```
Comment (dkalinin) [inline app/component/link.js]: <br> is not a responsibility of Link. Also, root element of Link can be <a>, there's no need to wrap with <span>.

File: app/component/login-form.js [hunk 3]
```diff
@@ -18,24 +19,60 @@ export class LoginForm {
...
-    form.appendChild(referenceToRegister);
+    new Link(this.form, 'Not registered?', '#/registration');
+    this.parentNode.appendChild(this.form);
+  }
+
+  /**
+   * Adds error message to form if something went wrong.
+   * @param {string} message
+   */
+  addErrorMessage(message) {
```
Comment (dkalinin) [inline app/component/login-form.js]: How can I remove error message?

File: app/component/registration-form.js [hunk 4]
```diff
@@ -19,47 +21,72 @@ export class RegistrationForm {
    * Renders and displays form object.
    */
   render() {
-    const form = document.createElement('form');
-    form.id = 'registration-form';
+    this.form = document.createElement('this.form');
+    this.form.id = 'registration-this.form';
```
Comment (dkalinin) [inline app/component/registration-form.js]: Let's not use periods in ids.

File: app/component/registration-form.js [hunk 5]
```diff
@@ -19,47 +21,72 @@ export class RegistrationForm {
...
+    errorMessage.innerHTML = `${message}`;
+    this.form.appendChild(errorMessage);
+  }
+
+  /**
+   *
+   * @param {Function} callback
+   */
+  onSubmit(callback) {
+    this._callback = callback;
```
Comment (dkalinin) [inline app/component/registration-form.js]: There's no need to save a callback to a property.

File: app/component/registration-form.js [hunk 6]
```diff
@@ -19,47 +21,72 @@ export class RegistrationForm {
...
-      login.errorField = loginValidationResult;
+      const loginValidationResult = FormValidator.validateLogin(this._login.value);
+      this._login.errorField = loginValidationResult;
 
-      const passwordValidationResult = FormValidator.validatePasswords(password.value,
-          secondPassword.value);
-      password.errorField = passwordValidationResult;
+      const passwordValidationResult = FormValidator.validatePasswords(this._password.value,
+          this._secondPassword.value);
+      this._password.errorField = passwordValidationResult;
```
Comment (dkalinin) [inline app/component/registration-form.js]: Please move all validation calls to a separate method.

File: app/controller/login-controller.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  /**
+   * Renders login form and sets `submit` event on it.
+   */
+  render() {
+    const loginForm = new LoginForm(this.parentNode);
+
+    loginForm.onSubmit(()=>{
+      RequestService.get(loginForm.userLogin, loginForm.userPassword)
+          .then(()=>Router.redirect('#/file-list'))
+          .catch(()=>loginForm.addErrorMessage('User already exists.'));
```
Comment (dkalinin) [inline app/controller/login-controller.js]: Is ESLint OK with no spaces between () and => ?

File: app/controller/registration-controller.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  /**
+   * Renders registration form and sets `submit` event on it.
+   */
+  render() {
+    const registrationForm = new RegistrationForm(this.parentNode);
+    registrationForm.onSubmit(()=>{
+      RequestService.post(registrationForm.user)
+          .then(()=>Router.redirect('#/login'))
```
Comment (dkalinin) [inline app/controller/registration-controller.js]: Let's have a router as an attribute of Controller. It should be passed from Application, where router is initialized.

File: app/service/request-service.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    });
+  }
+
+  /**
+   * Sends get request to BE to check if user with given login and password exists.
+   * @param {string} userLogin
+   * @param {string} userPassword
+   * @return {Promise<any>}
+   */
+  static get(userLogin, userPassword) {
```
Comment (dkalinin) [inline app/service/request-service.js]: `GET` request doesn't have a payload. So for logging in `POST` should be used.

Comment (denysLavrienov) [reply app/service/request-service.js]: It is written for further implementation.

File: app/controller/login-controller.js (hunk 7 above)
Comment (denysLavrienov) [reply app/controller/login-controller.js]: image

File: app/component/button.js [hunk 10]
```diff
@@ -29,6 +30,9 @@ export class Button {
    * @param {Function} handler function that is executed when button is clicked.
    */
   onClick(handler) {
-    this.button.addEventListener('click', handler);
+    if (!this.disabled) {
```
Comment (dkalinin) [inline app/component/button.js]: This is a responsibility of Form.

File: app/component/input-field.js [hunk 11]
```diff
@@ -65,8 +65,10 @@ export class InputField {
   /**
    * Creates event `change` for input field.
    * @param {Function} handler function that is executed when field is changed.
+   * @return {Object} this
```
Comment (dkalinin) [inline app/component/input-field.js]: Not Object but InputField.

File: app/component/login-form.js [hunk 12]
```diff
@@ -18,24 +19,69 @@ export class LoginForm {
...
+   */
+  get userLogin() {
+    return this._login.value;
+  }
+
+  /**
+   * @return {string}
+   */
+  get userPassword() {
+    return this._password;
```
Comment (dkalinin) [inline app/component/login-form.js]: Let's use event to pass data from this form.

File: app/component/registration-form.js [hunk 13]
```diff
@@ -19,47 +21,79 @@ export class RegistrationForm {
    * Renders and displays form object.
    */
   render() {
-    const form = document.createElement('form');
-    form.id = 'registration-form';
+    this.form = document.createElement('this-form');
```
Comment (dkalinin) [inline app/component/registration-form.js]: PLease fix.

**Pull request merged**